### PR TITLE
expression: check if period is valid in `period_add`

### LINKED ISSUE #9736: `PERIOD_ADD` is not compatible with Mysql
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
SELECT PERIOD_ADD(-11924, 2892462180);
```

2. What did you expect to see?
In Mysql 8.0.12
```
mysql> SELECT PERIOD_ADD(-11924, 2892462180);
ERROR 1210 (HY000): Incorrect arguments to period_add
```


3. What did you see instead?
In TiDB
```
mysql> SELECT PERIOD_ADD(-11924, 2892462180);
+--------------------------------+
| PERIOD_ADD(-11924, 2892462180) |
+--------------------------------+
|                    24103840192 |
+--------------------------------+
1 row in set (0.00 sec)
```



4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                     |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.0-beta-212-g0681b8125
Git Commit Hash: 0681b8125080045ee0d4747ce61923164c6f2dd1
Git Branch: master
UTC Build Time: 2019-03-14 04:39:19
GoVersion: go version go1.12 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
fix #9736.

### What is changed and how it works?
add `validPeriod`.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by mantuliu** [comment]
#10332  is the same question. @qw4990 @lonng

**Comment by qw4990** [comment]
> #10332 is the same question. @qw4990 @lonng

Yes, it seems that #10332 can be fixed by this PR.

File: expression/builtin_time.go [hunk 1]
```diff
@@ -4785,15 +4789,15 @@ func (b *builtinPeriodAddSig) evalInt(row chunk.Row) (int64, bool, error) {
...
-	if p == 0 {
-		return 0, false, nil
-	}
-
 	n, isNull, err := b.args[1].EvalInt(b.ctx, row)
 	if isNull || err != nil {
 		return 0, true, err
 	}
 
+	if !validPeriod(p) {
```
Comment (zz-jason) [inline expression/builtin_time.go]: should we validate this argument before `b.args[1].EvalInt(b.ctx, row)`?

Comment (qw4990) [reply expression/builtin_time.go]: No... in MySQL, if `p` is invalid but `n` is `NULL`, the result is `NULL`, so we have to check if `n` is `NULL` first.

Comment (XuHuaiyu) [reply expression/builtin_time.go]: Maybe we also need to add this explanation as a comment.

Comment (qw4990) [reply expression/builtin_time.go]: Comment is added.

File: expression/builtin_time.go [hunk 2]
```diff
@@ -4734,6 +4734,10 @@ func (c *periodAddFunctionClass) getFunction(ctx sessionctx.Context, args []Expr
 	return sig, nil
 }
 
+func validPeriod(p int64) bool {
```
[the remark this answers is not in the document]
Comment (qw4990) [reply expression/builtin_time.go]: Comment is added.

**Comment by qw4990** [comment]
PTAL

**Comment by qw4990** [comment]
To solve the CI problem, I `git rebase` this PR to master without any change, so please help me to approve this PR again~ @lamxTyler @XuHuaiyu

**Pull request merged**